pkg/kds/mux/session.go

```diff
@@ -89,21 +115,68 @@ func (s *session) PeerID() string {
 	return s.peerID
 }
 
-func (s *session) Done() <-chan struct{} {
-	return s.done
+func (s *session) Error() <-chan error {
+	return s.err
+}
+
+type sendItem struct {
+	msg     *mesh_proto.Message
+	errChan chan error
+}
+
+type bufferStream struct {
+	sendBuffer chan sendItem
+	recvBuffer chan *mesh_proto.Message
+
+	lock   sync.Mutex // Protects the write side of the buffer, not the read
+	closed bool
 }
 
-func (s *session) Error() error {
-	select {
-	case err := <-s.err:
-		return err
-	default:
-		return nil
+func newBufferStream() *bufferStream {
+	return &bufferStream{
+		sendBuffer: make(chan sendItem, 1),
+		recvBuffer: make(chan *mesh_proto.Message, 1),
 	}
 }
 
-func (s *session) close() {
-	close(s.done)
-	close(s.serverStream.requests)
-	close(s.clientStream.responses)
+func (k *bufferStream) put(message *mesh_proto.Message) {
+	k.recvBuffer <- message
+}
```

Comment:
Because this method is only called from `handleRecv` and that is the only place that calls `close()`, writing to the streamn without holding the lock is OK.

I experimented with locking this, but it makes tests timeout since it serialized WRT `Send()`. Probably a rwlock is the right choice (hold the write when setting the closed flag).